quickjs.c

```diff
@@ -7945,101 +8000,73 @@ static JSValue JS_GetPropertyValue(JSContext *ctx, JSValue this_obj,
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_int32(p->u.array.u.int16_ptr[idx]);
-        case JS_CLASS_UINT16_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_int32(p->u.array.u.uint16_ptr[idx]);
-        case JS_CLASS_INT32_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_int32(p->u.array.u.int32_ptr[idx]);
-        case JS_CLASS_UINT32_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_uint32(p->u.array.u.uint32_ptr[idx]);
-        case JS_CLASS_BIG_INT64_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return JS_NewBigInt64(ctx, p->u.array.u.int64_ptr[idx]);
-        case JS_CLASS_BIG_UINT64_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return JS_NewBigUint64(ctx, p->u.array.u.uint64_ptr[idx]);
-        case JS_CLASS_FLOAT32_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_float64(p->u.array.u.float_ptr[idx]);
-        case JS_CLASS_FLOAT64_ARRAY:
-            if (unlikely(idx >= p->u.array.count)) goto slow_path;
-            return js_float64(p->u.array.u.double_ptr[idx]);
-        default:
-            goto slow_path;
-        }
-    } else {
-    slow_path:
-        atom = JS_ValueToAtom(ctx, prop);
-        JS_FreeValue(ctx, prop);
-        if (unlikely(atom == JS_ATOM_NULL))
-            return JS_EXCEPTION;
-        ret = JS_GetProperty(ctx, this_obj, atom);
-        JS_FreeAtom(ctx, atom);
-        return ret;
+        JSObject *p = JS_VALUE_GET_OBJ(this_obj);
+        uint32_t idx = JS_VALUE_GET_INT(prop);
+        JSValue val;
+        /* fast path for array and typed array access */
+        if (js_get_fast_array_element(ctx, p, idx, &val))
+            return val;
     }
+    atom = JS_ValueToAtom(ctx, prop);
+    JS_FreeValue(ctx, prop);
+    if (unlikely(atom == JS_ATOM_NULL))
+        return JS_EXCEPTION;
+    ret = JS_GetProperty(ctx, this_obj, atom);
+    JS_FreeAtom(ctx, atom);
+    return ret;
 }
 
 JSValue JS_GetPropertyUint32(JSContext *ctx, JSValue this_obj,
                              uint32_t idx)
 {
-    return JS_GetPropertyValue(ctx, this_obj, js_uint32(idx));
+    return JS_GetPropertyInt64(ctx, this_obj, idx);
+}
+
+/* Check if an object has a property. Return value:
+   -1 for exception, *pval set to JS_EXCEPTION
+   TRUE if property exists, stored into *pval,
+   FALSE if property does not exist. *pval set to JS_UNDEFINED.
+ */
+static int JS_TryGetProperty(JSContext *ctx, JSValue obj, JSAtom prop, JSValue *pval)
+{
+    /* could optimize by combining the calls, but hardly worth it */
+    int res = JS_HasProperty(ctx, obj, prop);
+    if (res < 0) {
+        *pval = JS_EXCEPTION;
+        return res;
+    }
+    if (res == FALSE) {
+        *pval = JS_UNDEFINED;
+        return FALSE;
+    }
```

Comment:
I don't think this is correct, although the problem is probably more with JS_HasProperty.

`JS_GetProperty(obj, prop)` with obj="bar" and prop=1 returns "a", whereas JS_HasProperty claims the property doesn't exist.

In a similar vein, obj="foo" and prop="length" should produce 3, not property-not-found.

Reply:
I agree, `JS_HasProperty` ignores properties of values that are not objects, which is the semantics of ECMA **7.3.12 HasProperty ( O, P )** that requires **O** to be an Object.

Similarly, `JS_TryGetProperty()` and `JS_TryGetPropertyInt64()` must only be called for Objects. These APIs should probably take a `JSObject *` but changing that would have caused too many modifications.

Reply:
I am not sure `JS_TryGetProperty()` is useful after all.  Optimizing `JS_TryGetPropertyInt64()` is definitely a good idea for array methods, but a more general function does not seem to provide a real advantage and may create some confusion.  I shall investigate some more and post and updated patch.

Reply:
My main motivation for #105 is to avoid looking up the property twice. In that respect a JS_TryGetProperty that's just JS_HasProperty + JS_GetProperty without caching the lookup indeed isn't useful.

Reply:
A better approach to avoiding the double lookup is to use an alternate `JS_GetPropertyInternal` function that would update a structure argument with the location of the property found: prototype level, fast_array / shape entry, getter?, exotic method call...

Reply:
Yes, that's pretty much what I was thinking.

Reply:
OK, I shall simplify this PR and keep only the `JS_TryGetPropertyInt64`.